apps/src/javalab/JavalabView.jsx

```diff
@@ -107,6 +137,83 @@ class JavalabView extends React.Component {
     });
   };
 
+  handleWidthResize = desiredWidth => {
+    let newWidth = Math.max(100, Math.min(desiredWidth, 600));
+    this.props.setLeftWidth(newWidth);
+
+    this.updateLayoutThrottled(newWidth);
+  };
+
+  updateLayout = availableWidth => {
+    // We scale the visualization to take up as much space as it can, both
+    // vertically and horizontally.  Its width is constrained by the width
+    // of the left side, which the user can adjust by dragging a resizer, and
+    // which is passed into this function as availableWidth.
+    // Its height is constrained by how much of the window is available.  We
+    // start with the entire height of the window, subtract the space used by
+    // the instructions (which includes their header and horizontal resizer),
+    // subtract the space used by misc existing elements (which includes the
+    // page header, gaps between areas, the "preview" header, and the small
+    // footer at the bottom), and subtract the space used by the speed slider
+    // if it's shown.
+    // The visualization is a square, so the width it's rendered at will be
+    // constrained by both the available width and height.
+    const miscExistingElementsHeight = 135;
+    const sliderHeight = 60;
+
+    // The original visualization is rendered at 800x800.
+    const originalVisualizationWidth = 800;
+
+    // Determine the available height.
+    let availableHeight =
+      window.innerHeight -
+      this.props.topInstructionsHeight -
+      miscExistingElementsHeight;
+    if (this.props.viewMode === CsaViewMode.NEIGHBORHOOD) {
+      availableHeight -= sliderHeight;
+    }
+
+    // Use the biggest available size.
+    let newVisualizationWidth = Math.min(availableHeight, availableWidth);
+
+    // Scale the visualization.
+    let scale = newVisualizationWidth / originalVisualizationWidth;
+    if (scale < 0) {
+      // Avoid inverting.
+      scale = 0;
+    }
+    const scaleCss = `scale(${scale})`;
+    if (this.props.viewMode === CsaViewMode.NEIGHBORHOOD) {
+      $('#svgMaze').css('transform', scaleCss);
+    } else if (this.props.viewMode === CsaViewMode.THEATER) {
+      $('#theater-container').css('transform', scaleCss);
+    }
```

Comment:
I see that we have a few viewMode specific branches here - I wonder if it would be a good idea to split some of this logic out into functions (maybe like `computeAvailableHeightForViewMode(viewMode)`, `getContainerIdForViewMode(viewMode)`) so that we're hopefully less likely to forget to update this if/when we add a new viewMode?

Reply:
Good idea!  Also added to backlog.